### PR TITLE
Allow filtering metrics using labels

### PR DESCRIPTION
**Pull request by pierresouchay**
This will allow for a more elegant solution to
https://github.com/hashicorp/consul/pull/4042

TODO:

* [X] Add unit tests

### CODE REVIEW

File: start.go [hunk 1]
```diff
@@ -23,17 +23,21 @@ type Config struct {
...
-	lastNumGC  uint32
-	sink       MetricSink
-	filter     *iradix.Tree
-	filterLock sync.RWMutex
+	lastNumGC     uint32
+	sink          MetricSink
+	filter        *iradix.Tree
+	allowedLabels map[string]int
+	blockedLabels map[string]int
+	filterLock    sync.RWMutex
```
Comment (banks) [inline start.go]: nitpick: I'd comment that all of `filter, allowedLabels, blockedLabels` are protected by `filterLock` as that's not implied by the name (no need to rename IMO just worth clarifying).

File: metrics.go [hunk 2]
```diff
@@ -146,20 +172,54 @@ func (m *Metrics) UpdateFilter(allow, block []string) {
 	}
 }
 
+// labelIsAllowed return true if a label has to be included
```
[the remark this answers is not in the document]
Comment (pierresouchay) [reply metrics.go]: DONE in next patch

File: metrics.go [hunk 3]
```diff
@@ -146,20 +172,54 @@ func (m *Metrics) UpdateFilter(allow, block []string) {
...
+	if m.allowedLabels != nil {
+		_, ok := m.allowedLabels[labelName]
+		return ok
+	}
+	// Allow by default
+	return true
+}
+
+// filterLabels return only allowed labels
+func (m *Metrics) filterLabels(labels []Label) []Label {
```
[the remark this answers is not in the document]
Comment (pierresouchay) [reply metrics.go]: DONE in next patch

File: start.go (hunk 1 above)
Comment (pierresouchay) [reply start.go]: DONE in next patch

**Comment by pierresouchay** [comment]
@banks Suggestion to go forward with thoses patches ?

Kind Regards

**Comment by banks** [comment]
I think @kyhavlov had some suggestions he was going to make here. 

We’ve had a busy couple of weeks but I’ll ping him today to take a look.

**Comment by pierresouchay** [comment]
Hello @kyhavlov, @banks ,

Could someone have a look, so I can continue on https://github.com/hashicorp/consul/pull/4042 ?

Kind regards

**Comment by iksaif** [comment]
Looks good to me :+1:

**Comment by kyhavlov** [comment]
Hey @pierresouchay sorry for the delayed response. I think the idea behind this seems sound, but it feels weird to be blocking labels globally by name without it being attached to series name at all. Do you think it makes sense to make the label filters specific to a given series name for more fine-grained blocking?

**Comment by pierresouchay** [comment]
@kyhavlov Well for our use-case (aka adding labels in Consul everywhere), it will be more pain than ease, on a API point of view, it make more sense.

What would you like me to do ?

* Regex based ? (In that case, it probably means that all calls with pay the price for matching regex)
* prefix ? (I that case the patch will be quite invasive since I will have to change all radix based structures)

Tell me what you want in that case.

Kind regards

File: start.go [hunk 4]
```diff
@@ -23,17 +23,21 @@ type Config struct {
 
 	AllowedPrefixes []string // A list of metric prefixes to allow, with '.' as the separator
 	BlockedPrefixes []string // A list of metric prefixes to block, with '.' as the separator
+	AllowedLabels   []string // A list of metric labels to allow, with '.' as the separator
+	BlockedLabels   []string // A list of metric prefixes to block, with '.' as the separator
```
[the remark this answers is not in the document]
Comment (pierresouchay) [reply start.go]: DONE

File: start.go [hunk 5]
```diff
@@ -23,17 +23,21 @@ type Config struct {
...
 type Metrics struct {
 	Config
-	lastNumGC  uint32
-	sink       MetricSink
-	filter     *iradix.Tree
-	filterLock sync.RWMutex
+	lastNumGC     uint32
+	sink          MetricSink
+	filter        *iradix.Tree
+	allowedLabels map[string]int
```
[the remark this answers is not in the document]
Comment (pierresouchay) [reply start.go]: DONE

**Comment by pierresouchay** [comment]
@kyhavlov Thank you for the review, PR updated with requested fixes

**Comment by pierresouchay** [comment]
Hello @kyhavlov @banks ,

I made the fixes, would you consider merging it so I can go forward on https://github.com/hashicorp/consul/pull/4042 ?

Thanks a lot

Kind regards

File: metrics.go [hunk 6]
```diff
@@ -120,23 +124,45 @@ func (m *Metrics) MeasureSinceWithLabels(key []string, start time.Time, labels [
...
-	m.sink.AddSampleWithLabels(key, msec, labels)
+	m.sink.AddSampleWithLabels(key, msec, labelsFiltered)
 }
 
 // UpdateFilter overwrites the existing filter with the given rules.
 func (m *Metrics) UpdateFilter(allow, block []string) {
+	m.UpdateFilterAndLabels(allow, block, m.AllowedLabels, m.BlockedLabels)
+}
+
+// UpdateFilterAndLabels overwrites the existing filter with the given rules.
```
Comment (banks) [inline metrics.go]: Can we make this doc comment explicit what the behaviour of label filter is? i.g. apply globally to all metrics and apply orthogonally to the allow/block prefixes. I'd also include the use-case for them (globally configuring whether potentially high-cardinality metrics are included) to make it obvious why they work like this.

I think that satisfied the last point Kyle mentioned.

Comment (pierresouchay) [reply metrics.go]: @banks DONE: I added more text in README.md as well

**Pull request merged**